File: workshop2/php/Manazer.php
<?php

class Manazer
{
    public function __construct(private IDownloader $downloader, private IParser $parser)
    {
    }

    public function nactiData(): SeznamZviratek
    {
        $data = $this->downloader->download(__DIR__ . '/../zviratka.csv');
        return $this->parser->process($data);
    }

    public function nejlevnejsiZviratko(SeznamZviratek $zviratka): string
    {
        // Cílem úpravy této metody je použít pouze následující řádek:
        //return $zviratka->najdiNejlevnejsiZviratko()->nazev;


        $nejlevnejsi = null;
        $nazev = '';
        foreach ($zviratka->zviratka as $zviratko) {
            if ($nejlevnejsi === null) {
                $nejlevnejsi = $zviratko[4];
                $nazev = $zviratko[1];
                continue;
            }

            if ($zviratko[4] < $nejlevnejsi) {
                $nejlevnejsi = $zviratko[4];
                $nazev = $zviratko[1];
            }
        }

        var_dump(sprintf('Nejlevnější cena: %d', $nejlevnejsi));
        return $nazev;
    }

    public function najdiZviratkoPodleNazvu(SeznamZviratek $zviratka, string $nazev): Zviratko|null
    {
        return $zviratka->najdiZviratkoPodleNazvu($nazev);
    }
}

File: workshop2/python/manazer.py
from pprint import pprint
from typing import Union

from downloader import Downloader
from parser import Parser
from model import SeznamZviratek, Zviratko


class Manazer:

    def __init__(self, downloader: Downloader, parser: Parser):
        self.downloader = downloader
        self.parser = parser

    def nacti_data(self) -> SeznamZviratek:
        return self.parser.process(self.downloader.get_file('../zviratka.csv'))

    def nejlevnejsi_zviratko(self, seznam_zviratek: SeznamZviratek) -> str:
        # Cílem úpravy této metody je použít pouze následující řádek:
        # return seznam_zviratek.najdi_nejlevnejsi_zviratko().nazev;

        nejlevnejsi = None
        nazev = ''

        for zviratko in seznam_zviratek.zviratka:
            if nejlevnejsi is None:
                nejlevnejsi = int(zviratko[4])
                nazev = zviratko[1]
                continue

            if int(zviratko[4]) < int(nejlevnejsi):
                nejlevnejsi = zviratko[4]
                nazev = zviratko[1]

        pprint(f'Nejlevnější cena: {nejlevnejsi}')

        return nazev

    def najdi_zviratko_podle_nazvu(self, seznam_zviratek: SeznamZviratek, nazev: str) -> Union[Zviratko, None]:
        return seznam_zviratek.najdi_zviratko_podle_nazvu(nazev)
